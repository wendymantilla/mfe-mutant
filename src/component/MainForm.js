import React, {useEffect, useState} from 'react';
import DnaButtons from './DnaButtons';
import {getReport, isMutant} from '../service/apiService'
import DnaTable from "./DnaTable";
import ReportTable from "./ReportTable";
import Alert from "./Alert";
import ReportCountTable from "./ReportCountTable";

const MainForm = () => {

        const [formData, setFormData] = useState({name: '', age: '', dna: []});

        const [alert, setAlert] = useState({show: true, type: "info", message: "Enter the DNA sequence"});

        const [reportData, setReportData] = useState({})

        const [dnaData, setDnaData] = useState("")

        const handleChange = (e) => {
            const {name, value} = e.target;
            setFormData({
                ...formData,
                [name]: value
            });
        };

        const handleDNAButtonClick = (letter) => {
            if (formData.dna.length >= 6) {
                setAlert({...alert, message: "DNA is complete!", type: "warning", show: true})
                return
            }
            if (dnaData.length < 6) {
                setDnaData(dnaData.concat(letter))
            } else {
                setFormData({
                    ...formData,
                    dna: [...formData.dna, dnaData]
                });
                setDnaData("")
            }
        };

        const handleSubmit = (e) => {
            e.preventDefault();
            if (formData.dna.length >= 6) {
                console.log(formData);
                isMutant(formData)
                    .then(data => setAlert({
                        ...alert,
                        message: data.is_mutant ? `${formData.name} is a mutant!` : `${formData.name} is not a mutant!`,
                        type: "success",
                        show: true
                    }))
                    .catch((err) => setAlert({...alert, message: err.message, type: "error", show: true}))
                //handleReset(e);

            } else {
                setAlert({...alert, message: "DNA is NOT complete!", type: "error", show: true})
            }
        };

        const handleReset = (e) => {
            e.preventDefault();
            setDnaData('')
            setFormData({name: '', age: '', dna: []})
            setAlert({show: true, type: "info", message: "Enter the DNA sequence"})
        }


        useEffect(() => {
            getReport()
                .then(response => setReportData(response))
                .catch(err => setAlert({...alert, message: err, type: "error", show: true}))
        }, [alert])

        const generateDNA = () => {
            const letters = ['A', 'C', 'G', 'T'];
            const random = Array.from({length: 6}, () => Array.from({length: 6}, () => letters[Math.floor(Math.random() * letters.length)]).join(''))
            setFormData({...formData, dna: random})
        }

        return (
            <div className="card">
                <div className="card-header">
                    <h2 className="card-title">Mutant Test</h2>
                </div>
                <div className="card-body">
                    <div className="row">
                        <div className="col-7 mt-3">
                            <form onSubmit={handleSubmit} onReset={handleReset} className="form-inline">
                                <div className="input-group mb-3">
                                    <span className="input-group-text">Name:</span>
                                    <input
                                        type="text"
                                        className="form-control"
                                        id="name"
                                        name="name"
                                        title="name"
                                        value={formData.name}
                                        onChange={handleChange}
                                        maxLength="20"
                                        required
                                    />
                                </div>
                                <div className="input-group mb-3">
                                    <span className="input-group-text">Age:</span>
                                    <input
                                        type="number"
                                        className="form-control"
                                        id="age"
                                        name="age"
                                        title="age"
                                        value={formData.age}
                                        maxLength="3"
                                        onChange={handleChange}
                                        required
                                    />
                                </div>
                                <DnaButtons handleButtonClick={handleDNAButtonClick} data={dnaData}
                                            generateDNA={generateDNA}/>
                                <Alert alert={alert}/>

                                <div className="btn-group">
                                    <button type="submit" className="btn btn-outline-primary ">Send</button>
                                    <button type="reset" className="btn btn-outline-info ">Clean</button>
                                </div>
                            </form>
                            <ReportCountTable reportData={reportData}/>
                        </div>
                        <div className="col">
                            <DnaTable dna={formData.dna}/>
                        </div>
                    </div>
                    <ReportTable reportData={reportData}/>
                </div>

            </div>


        );
    }
;

export default MainForm;
