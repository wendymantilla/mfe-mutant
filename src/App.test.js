import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import MainForm from "./component/MainForm";

test('renders MainForm component', () => {
  const { getByText, getByLabelText } = render(<MainForm />);

  expect(getByText('Mutant Test')).toBeInTheDocument();

  expect(getByText('Name:')).toBeInTheDocument();
  expect(getByText('Age:')).toBeInTheDocument();

  expect(getByText('Send')).toBeInTheDocument();
  expect(getByText('Clean')).toBeInTheDocument();
});