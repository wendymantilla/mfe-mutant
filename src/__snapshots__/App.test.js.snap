// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`changes the class when hovered 1`] = `
<div
  className="container mt-4"
>
  <div
    className="card"
  >
    <div
      className="card-header"
    >
      <h2
        className="card-title"
        href="/"
      >
        Mutant Test
      </h2>
    </div>
    <div
      className="card-body"
    >
      <div
        className="row"
      >
        <div
          className="col-7 mt-3"
        >
          <form
            className="form-inline"
            onReset={[Function]}
            onSubmit={[Function]}
          >
            <div
              className="input-group mb-3"
            >
              <span
                className="input-group-text"
                htmlFor="name"
              >
                Name:
              </span>
              <input
                className="form-control"
                id="name"
                maxLength="20"
                name="name"
                onChange={[Function]}
                required={true}
                type="text"
                value=""
              />
            </div>
            <div
              className="input-group mb-3"
            >
              <span
                className="input-group-text"
              >
                Age:
              </span>
              <input
                className="form-control"
                id="age"
                maxLength="3"
                name="age"
                onChange={[Function]}
                required={true}
                type="number"
                value=""
              />
            </div>
            <div
              className="input-group mb-3"
            >
              <span
                className="input-group-text"
              >
                DNA:
              </span>
              <div
                aria-label="Botones DNA"
                className="btn-group"
                role="group"
              >
                <button
                  className="btn btn-outline-primary  "
                  onClick={[Function]}
                  type="button"
                >
                  <span
                    className="p-2 font-monospace"
                  >
                    A
                  </span>
                </button>
                <button
                  className="btn btn-outline-primary  "
                  onClick={[Function]}
                  type="button"
                >
                  <span
                    className="p-2 font-monospace"
                  >
                    C
                  </span>
                </button>
                <button
                  className="btn btn-outline-primary  "
                  onClick={[Function]}
                  type="button"
                >
                  <span
                    className="p-2 font-monospace"
                  >
                    G
                  </span>
                </button>
                <button
                  className="btn btn-outline-primary  "
                  onClick={[Function]}
                  type="button"
                >
                  <span
                    className="p-2 font-monospace"
                  >
                    T
                  </span>
                </button>
                <button
                  className="btn btn-outline-success "
                  onClick={[Function]}
                  type="button"
                >
                  <span
                    className="p-2 font-monospace"
                  >
                    AUTO
                  </span>
                </button>
              </div>
              <input
                className="form-control"
                id="dna"
                name="dna"
                readOnly={true}
                type="text"
                value=""
              />
            </div>
            <div
              className="alert mb-3 alert-primary"
              role="alert"
              style={
                Object {
                  "display": "block",
                }
              }
            >
              Enter the DNA sequence
            </div>
            <div
              className="btn-group"
            >
              <button
                className="btn btn-outline-primary "
                type="submit"
              >
                Send
              </button>
              <button
                className="btn btn-outline-info "
                type="reset"
              >
                Clean
              </button>
            </div>
          </form>
          <div
            className="table-responsive p-3"
          >
            <table
              className="table table-striped table-sm table-bordered"
            >
              <thead>
                <tr>
                  <th
                    scope="col"
                  >
                    Human DNA
                  </th>
                  <th
                    scope="col"
                  >
                    Mutant DNA
                  </th>
                  <th
                    scope="col"
                  >
                    Mutant ratio
                  </th>
                </tr>
              </thead>
              <tbody>
                <tr>
                  <td />
                  <td />
                  <td>
                    NaN
                  </td>
                </tr>
              </tbody>
            </table>
          </div>
        </div>
        <div
          className="col"
        >
          <div
            className="input-group "
          >
            <ul
              className="list-group list-group-flush mx-auto "
            />
          </div>
        </div>
      </div>
      <div
        className="table-responsive p-3"
      >
        <table
          className="table table-striped table-sm table-bordered"
        >
          <thead>
            <tr>
              <th
                scope="col"
              >
                #
              </th>
              <th
                scope="col"
              >
                Name
              </th>
              <th
                scope="col"
              >
                Age
              </th>
              <th
                scope="col"
              >
                Is mutant
              </th>
            </tr>
          </thead>
          <tbody />
        </table>
      </div>
    </div>
  </div>
</div>
`;
